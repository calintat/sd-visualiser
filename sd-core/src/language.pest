variable = @{ !("bind" | "in") ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }
bind_clause = { "bind" ~ variable ~ "=" ~ term ~ "in" }
expr = { bind_clause* ~ value }
term = { thunk | (active_op ~ "(" ~ args ~ ")") | value }
value = { (passive_op ~ "(" ~ args ~ ")") | variable }
active_op = { "plus" | "minus" | "times" | "eq" | "if" | "app" | "lambda" | "rec" }
passive_op = { number }
number = _{ ASCII_DIGIT+ }
args = _{ value ~ ("," ~ value)* | "" }
program = _{ SOI ~ expr ~ EOI }
WHITESPACE = _{ " " | "\t" | "\n" }
thunk = { variable* ~ "." ~ expr }