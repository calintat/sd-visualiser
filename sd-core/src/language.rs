pub mod chil;
pub mod spartan;

pub(super) mod visitor;
