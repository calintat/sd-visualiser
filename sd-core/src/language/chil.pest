program = _{ SOI ~ expr ~ EOI }

keyword = { "def" | "output" | "thunk" }

expr = { bind_clause* ~ "output" ~ value }
bind_clause = { "def" ~ variable_def ~ "=" ~ value }

thunk = { "thunk" ~ addr ~ "=" ~ "{" ~ (variable_def ~ ("," ~ variable_def)*)? ~ "=>" ~ expr ~ "}" }

value = { variable | op ~ ("(" ~ (thunk+ | values ~ (";" ~ thunk+)?) ~ ")")? }
values = _{ value ~ ("," ~ value)* }

op = ${ !keyword ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "+" | "-" | "*" | "=")* ~ ("/" ~ (number | string | ty ~ "/" ~ ty))? }
number = @{ ASCII_DIGIT+ ~ ("." ~ ASCII_DIGIT+)? }
string = { "\"" ~ (!"\"" ~ ANY)* ~ "\"" }

ty = { generic_ty | base_ty | function_ty | tuple_ty }
tys = _{ ty ~ ("," ~ ty)* }
base_ty = @{ 'A'..'Z' ~ (ASCII_ALPHANUMERIC | "-")* }
generic_ty = { base_ty ~ "<" ~ tys ~ ">" }
tuple_ty = { "(" ~ tys? ~ ")" }
function_ty = { tuple_ty ~ "->" ~ ty }

variable = { addr | identifier ~ "(" ~ "id" ~ ":" ~ addr ~ ")" }
variable_def = { variable ~ ":" ~ ty | variable }

COMMENT = _{ "#" ~ (!"\n" ~ ANY)* }
WHITESPACE = _{ " " | "\t" | "\n" }

addr = @{ ("@" | "%") ~ ASCII_DIGIT+ }
identifier = @{ !keyword ~ "_"* ~ ASCII_ALPHA ~ (ASCII_ALPHANUMERIC | "_")* }