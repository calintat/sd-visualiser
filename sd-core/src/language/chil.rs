#![allow(clippy::clone_on_copy)]

use std::fmt::{Display, Write};

use from_pest::{ConversionError, FromPest, Void};
use pest::iterators::Pairs;
use pest_ast::FromPest;
use pest_derive::Parser;

use super::{span_into_str, AsVar};

pub struct Chil;

impl super::Language for Chil {
    type Op = Op;
    type Var = Variable;
    type Addr = Addr;
    type VarDef = VariableDef;

    type Rule = Rule;

    fn expr_rule() -> Self::Rule {
        Rule::expr
    }
    fn bind_rule() -> Self::Rule {
        Rule::bind
    }
    fn value_rule() -> Self::Rule {
        Rule::value
    }
    fn thunk_rule() -> Self::Rule {
        Rule::thunk
    }
}

pub type Expr = super::Expr<Chil>;
pub type Bind = super::Bind<Chil>;
pub type Value = super::Value<Chil>;
pub type Thunk = super::Thunk<Chil>;

#[derive(Parser)]
#[grammar = "language/chil.pest"]
pub struct ChilParser;

fn parse_addr_first(input: &str) -> char {
    input.chars().next().unwrap()
}

fn parse_addr_second(input: &str) -> usize {
    input[1..].parse().unwrap()
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Op(pub String);

impl Display for Op {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0.as_str() {
            "+" | "throwing+" => f.write_char('+'),
            "-" | "throwing-" => f.write_char('-'),
            "*" | "throwing*" => f.write_char('*'),
            "/" | "throwing/" => f.write_char('/'),
            "%" | "throwing%" => f.write_char('%'),
            "&&" | "throwing&&" => f.write_str("&&"),
            "||" | "throwing||" => f.write_str("||"),
            "!" | "throwing!" => f.write_char('!'),
            "==" | "throwing==" => f.write_char('='),
            "!=" | "throwing!=" => f.write_char('≠'),
            "<" | "throwing<" => f.write_char('<'),
            "<=" | "throwing<=" => f.write_char('≤'),
            ">" | "throwing>" => f.write_char('>'),
            ">=" | "throwing>=" => f.write_char('≥'),
            "func" => f.write_char('λ'),
            "seq" => f.write_char(';'),
            "unit" => f.write_str("()"),
            str => {
                if str.starts_with("apply/") {
                    return f.write_char('@');
                }
                if str.starts_with("tuple/") {
                    return f.write_char('⊗');
                }
                if let Some(rest) = None
                    .or_else(|| str.strip_prefix("bool/"))
                    .or_else(|| str.strip_prefix("int64/"))
                    .or_else(|| str.strip_prefix("float64/"))
                    .or_else(|| str.strip_prefix("string/"))
                {
                    return f.write_str(rest);
                }
                f.write_str(str)
            }
        }
    }
}

impl<'pest> FromPest<'pest> for Op {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::op {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable))]
pub struct Variable {
    pub name: Option<Identifier>,
    pub addr: Addr,
}

impl Display for Variable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.name {
            None => write!(f, "{}", self.addr),
            Some(name) => write!(f, "{}(id: {})", name, self.addr),
        }
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::addr))]
pub struct Addr(
    #[pest_ast(outer(with(span_into_str), with(parse_addr_first)))] pub char,
    #[pest_ast(outer(with(span_into_str), with(parse_addr_second)))] pub usize,
);

impl Display for Addr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", self.0, self.1)
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct Identifier(pub String);

impl Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl<'pest> FromPest<'pest> for Identifier {
    type Rule = Rule;
    type FatalError = Void;

    fn from_pest(
        pest: &mut Pairs<'pest, Self::Rule>,
    ) -> Result<Self, ConversionError<Self::FatalError>> {
        let mut clone = pest.clone();
        let pair = clone.next().ok_or(ConversionError::NoMatch)?;
        if pair.as_rule() != Rule::identifier {
            return Err(ConversionError::NoMatch);
        }
        *pest = clone;
        Ok(Self(pair.as_str().to_owned()))
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::variable_def))]
pub struct VariableDef {
    pub var: Variable,
    pub r#type: Option<Type>,
}

impl AsVar<Variable> for VariableDef {
    fn as_var(&self) -> &Variable {
        &self.var
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::ty))]
pub enum Type {
    Base(BaseType),
    Generic(GenericType),
    Tuple(TupleType),
    Function(FunctionType),
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::base_ty))]
pub struct BaseType(#[pest_ast(outer(with(span_into_str), with(str::to_string)))] pub String);

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::generic_ty))]
pub struct GenericType {
    pub base: BaseType,
    pub params: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::tuple_ty))]
pub struct TupleType {
    pub types: Vec<Type>,
}

#[derive(Clone, Eq, PartialEq, Hash, Debug, FromPest)]
#[pest_ast(rule(Rule::function_ty))]
pub struct FunctionType {
    pub domain: TupleType,
    pub codomain: Box<Type>,
}

// Conversion to spartan

#[cfg(test)]
impl From<Op> for super::spartan::Op {
    fn from(op: Op) -> Self {
        match op.0.as_str() {
            "+" | "throwing+" => return Self::Plus,
            "-" | "throwing-" => return Self::Minus,
            "*" | "throwing*" => return Self::Times,
            "/" | "throwing/" => return Self::Div,
            "%" | "throwing%" => return Self::Rem,
            "&&" | "throwing&&" => return Self::And,
            "||" | "throwing||" => return Self::Or,
            "!" | "throwing!" => return Self::Not,
            "==" | "throwing==" => return Self::Eq,
            "!=" | "throwing!=" => return Self::Neq,
            "<" | "throwing<" => return Self::Lt,
            "<=" | "throwing<=" => return Self::Leq,
            ">" | "throwing>" => return Self::Gt,
            ">=" | "throwing>=" => return Self::Geq,
            "func" => return Self::Lambda,
            "atom" => return Self::Atom,
            "deref" => return Self::Deref,
            "asg" => return Self::Assign,
            "bool/true" => return Self::Bool(true),
            "bool/false" => return Self::Bool(false),
            str => {
                if str.starts_with("apply/") {
                    return Self::App;
                }
                if let Some(rest) = str.strip_prefix("int64/") {
                    return Self::Number(rest.parse().unwrap_or_default());
                }
                Self::Plus // dummy placeholder
            }
        }
    }
}

#[cfg(test)]
impl From<Variable> for super::spartan::Variable {
    fn from(var: Variable) -> Self {
        Self(format!("var_{}", var.addr.1))
    }
}

#[cfg(test)]
impl From<Addr> for super::spartan::Addr {
    fn from(_addr: Addr) -> Self {
        Self
    }
}

#[cfg(test)]
impl From<VariableDef> for super::spartan::Variable {
    fn from(def: VariableDef) -> Self {
        def.var.into()
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::path::Path;

    use dir_test::{dir_test, Fixture};
    use from_pest::FromPest;
    use pest::Parser;

    use super::{ChilParser, Expr, Rule};

    pub fn parse_chil(raw_path: &str) -> (&str, Expr) {
        let path = Path::new(raw_path);
        let program = std::fs::read_to_string(path).unwrap();
        let mut pairs = ChilParser::parse(Rule::program, &program).unwrap_or_else(|err| {
            panic!(
                "could not parse program {:?}\n{err:?}",
                path.file_stem().unwrap()
            )
        });
        let name = path.file_stem().unwrap().to_str().unwrap();
        let expr = Expr::from_pest(&mut pairs).unwrap();
        (name, expr)
    }

    #[allow(clippy::needless_pass_by_value)]
    #[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/*.chil", loader: crate::language::chil::tests::parse_chil, postfix: "check_parse")]
    fn check_parse(fixture: Fixture<(&str, Expr)>) {
        let (_name, _expr) = fixture.content();
    }
}
