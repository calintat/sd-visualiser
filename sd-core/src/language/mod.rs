use std::{
    fmt::{Debug, Display},
    hash::Hash,
};

use derivative::Derivative;

use crate::{common::Matchable, prettyprinter::PrettyPrint};

pub mod chil;
pub mod spartan;

pub(crate) fn span_into_str(span: pest::Span) -> &str {
    span.as_str()
}

pub trait GetVar<V> {
    fn var(&self) -> &V;
    fn into_var(self) -> V;
}

impl<V> GetVar<V> for V {
    fn var(&self) -> &V {
        self
    }

    fn into_var(self) -> V {
        self
    }
}

pub trait Fresh {
    fn fresh(number: usize) -> Self;
}

/// `Display` should give the symbolic representation (e.g. "+").
/// `PrettyPrint` should give the textual representation (e.g. "plus").
pub trait Syntax:
    Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint
{
}
impl<T: Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint> Syntax for T {}

pub struct BlockWithArgs<T: Language + ?Sized> {
    pub block: T::BlockAddr,
    pub args: Vec<Value<T>>,
}

pub enum CF<T: Language + ?Sized> {
    Return,
    Brs(Vec<BlockWithArgs<T>>),
}

pub trait ControlFlow<T: Language + ?Sized> {
    fn get_cf(&self) -> Option<CF<T>>;

    #[must_use]
    fn get_apply() -> Option<T::Op> {
        None
    }
}

pub trait Language {
    type Op: Syntax + ControlFlow<Self>;
    type Var: Syntax + Fresh;
    type Addr: Syntax;
    type BlockAddr: Syntax;
    type VarDef: Syntax + GetVar<Self::Var>;
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Expr<T: Language + ?Sized> {
    pub binds: Vec<Bind<T>>,
    pub values: Vec<Value<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Bind<T: Language + ?Sized> {
    pub defs: Vec<T::VarDef>,
    pub value: Value<T>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub enum Value<T: Language + ?Sized> {
    Variable(T::Var),
    Thunk(Thunk<T>),
    Op { op: T::Op, args: Vec<Value<T>> },
}

impl<T: Language> ControlFlow<T> for Value<T> {
    fn get_cf(&self) -> Option<CF<T>> {
        match self {
            Value::Op { op, .. } => op.get_cf(),
            _ => None,
        }
    }
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Thunk<T: Language + ?Sized> {
    pub addr: T::Addr,
    pub args: Vec<T::VarDef>,
    pub body: Expr<T>,
    pub blocks: Vec<Block<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Block<T: Language + ?Sized> {
    pub addr: T::BlockAddr,
    pub args: Vec<T::VarDef>,
    pub expr: Expr<T>,
}

// Conversions between languages

impl<T: Language> Expr<T> {
    pub fn into<U: Language>(self) -> Expr<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Expr {
            binds: self.binds.into_iter().map(Bind::into).collect(),
            values: self.values.into_iter().map(Value::into).collect(),
        }
    }
}

impl<T: Language> Bind<T> {
    pub fn into<U: Language>(self) -> Bind<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Bind {
            defs: self.defs.into_iter().map(Into::into).collect(),
            value: self.value.into(),
        }
    }
}

impl<T: Language> Value<T> {
    pub fn into<U: Language>(self) -> Value<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        match self {
            Self::Variable(var) => Value::Variable(var.into()),
            Self::Thunk(thunk) => Value::Thunk(thunk.into()),
            Self::Op { op, args } => Value::Op {
                op: op.into(),
                args: args.into_iter().map(Value::into).collect(),
            },
        }
    }
}

impl<T: Language> Thunk<T> {
    pub fn into<U: Language>(self) -> Thunk<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Thunk {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            body: self.body.into(),
            blocks: self.blocks.into_iter().map(Block::into).collect(),
        }
    }
}

impl<T: Language> Block<T> {
    pub fn into<U: Language>(self) -> Block<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Block {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            expr: self.expr.into(),
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{ffi::OsStr, path::Path};

    use super::{
        chil::tests::parse_chil,
        spartan::{self, tests::parse_sd},
    };

    pub fn parse(raw_path: &str) -> (&str, &str, spartan::Expr) {
        let path = Path::new(raw_path);
        match path.extension() {
            Some(ext) if ext == OsStr::new("sd") => {
                let (name, expr) = parse_sd(raw_path);
                ("sd", name, expr)
            }
            Some(ext) if ext == OsStr::new("chil") => {
                let (name, expr) = parse_chil(raw_path);
                ("chil", name, expr.into())
            }
            _ => unreachable!(),
        }
    }
}
