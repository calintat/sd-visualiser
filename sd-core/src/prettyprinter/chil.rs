use super::PrettyPrint;
use crate::language::chil::{
    Arg, AtomicType, BindClause, Expr, FunctionType, Identifier, Name, Op, Thunk, Type, Value,
    Variable, VariableDef,
};
use pretty::RcDoc;

impl PrettyPrint for Expr {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::concat(self.binds.iter().map(PrettyPrint::to_doc))
            .append(RcDoc::text("output"))
            .append(RcDoc::space())
            .append(self.value.to_doc())
    }
}

impl PrettyPrint for BindClause {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("def")
            .append(RcDoc::space())
            .append(self.var.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("="))
            .append(RcDoc::space())
            .append(self.value.to_doc())
            .append(RcDoc::line())
    }
}

impl PrettyPrint for Variable {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Id(id) => id.to_doc(),
            Self::Name(name, id) => name
                .to_doc()
                .append(RcDoc::text("("))
                .append(RcDoc::text("id"))
                .append(RcDoc::text(":"))
                .append(RcDoc::space())
                .append(id.to_doc())
                .append(RcDoc::text(")")),
        }
    }
}

impl PrettyPrint for VariableDef {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Inferred(var) => var.to_doc(),
            Self::Manifest { var, ty } => var
                .to_doc()
                .append(RcDoc::space())
                .append(RcDoc::text(":"))
                .append(RcDoc::space())
                .append(ty.to_doc()),
        }
    }
}

impl PrettyPrint for Type {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Atomic(aty) => aty.to_doc(),
            Self::Function(fty) => fty.to_doc(),
        }
    }
}

impl PrettyPrint for AtomicType {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(&self.0)
    }
}

impl PrettyPrint for FunctionType {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("(")
            .append(RcDoc::intersperse(
                self.ins.iter().map(PrettyPrint::to_doc),
                RcDoc::text(",").append(RcDoc::space()),
            ))
            .append(RcDoc::text(")"))
            .append(RcDoc::space())
            .append(RcDoc::text("->"))
            .append(RcDoc::space())
            .append(self.out.to_doc())
    }
}

impl PrettyPrint for Value {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Var(x) => x.to_doc(),
            Self::Op(op, args) => {
                if args.is_empty() {
                    op.to_doc()
                } else {
                    op.to_doc()
                        .append(RcDoc::text("("))
                        .append(RcDoc::intersperse(
                            args.iter().map(PrettyPrint::to_doc),
                            RcDoc::text(",").append(RcDoc::space()),
                        ))
                        .append(RcDoc::text(")"))
                }
            }
        }
    }
}

impl PrettyPrint for Thunk {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text("thunk")
            .append(RcDoc::space())
            .append(self.id.to_doc())
            .append(RcDoc::space())
            .append(RcDoc::text("="))
            .append(RcDoc::space())
            .append(RcDoc::text("{"))
            .append(RcDoc::space())
            .append(RcDoc::intersperse(
                self.args.iter().map(PrettyPrint::to_doc),
                RcDoc::text(",").append(RcDoc::space()),
            ))
            .append(RcDoc::space())
            .append(RcDoc::text("=>"))
            .append(RcDoc::line())
            .append(self.body.to_doc())
            .nest(4)
            .append(RcDoc::line())
            .append(RcDoc::text("}"))
    }
}

impl PrettyPrint for Arg {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        match self {
            Self::Value(v) => v.to_doc(),
            Self::Thunk(d) => d.to_doc(),
        }
    }
}

impl PrettyPrint for Op {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(&self.0)
    }
}

impl PrettyPrint for Name {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(&self.0)
    }
}

impl PrettyPrint for Identifier {
    fn to_doc(&self) -> RcDoc<'_, ()> {
        RcDoc::text(format!("%{}", self.0))
    }
}
