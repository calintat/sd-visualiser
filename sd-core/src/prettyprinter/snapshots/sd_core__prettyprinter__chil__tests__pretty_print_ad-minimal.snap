---
source: sd-core/src/prettyprinter/chil.rs
expression: expr.to_pretty()
---
def goo2(id: @61) : (Float64) -> Float64 = func(
  thunk %55 = { x(id: %62) : Float64 =>
    def %56 : Float64 = null
    def %57 : Float64 = atom(%56)
    def %60 : Float64 = block(%57;
      thunk %58 = { %59 : Float64 =>
        def %63 : Float64 = float64/0.000000
        def %64 : Bool = throwing>(x(id: %62), %63)
        def %188 : Bool = typeCast/Bool/Bool(%64)
        def %69 : Float64 = if(%188;
          thunk %65 = {  =>
            def %66 : Float64 = apply/1(goo1(id: @44), x(id: %62))
            def %190 : Float64 = typeCast/Float64/Float64(%66)
            output %190
          }
          thunk %67 = {  =>
            def %68 : Float64 = apply/1(hi(id: @34), x(id: %62))
            def %191 : Float64 = typeCast/Float64/Float64(%68)
            output %191
          }
        )
        def %70 : Nothing = break(%59, %69)
        output %70
      }
    )
    def %192 : Float64 = typeCast/Float64/Float64(%60)
    output %192
  }
)
def foo2(id: @94) : (Float64) -> Float64 = func(
  thunk %88 = { x(id: %95) : Float64 =>
    def %89 : Float64 = null
    def %90 : Float64 = atom(%89)
    def %93 : Float64 = block(%90;
      thunk %91 = { %92 : Float64 =>
        def %96 : Float64 = float64/0.000000
        def %97 : Bool = throwing>(x(id: %95), %96)
        def %179 : Bool = typeCast/Bool/Bool(%97)
        def %102 : Float64 = if(%179;
          thunk %98 = {  =>
            def %99 : Float64 = apply/1(foo1(id: @77), x(id: %95))
            def %181 : Float64 = typeCast/Float64/Float64(%99)
            output %181
          }
          thunk %100 = {  =>
            def %101 : Float64 = apply/1(hi(id: @34), x(id: %95))
            def %182 : Float64 = typeCast/Float64/Float64(%101)
            output %182
          }
        )
        def %103 : Nothing = break(%92, %102)
        output %103
      }
    )
    def %183 : Float64 = typeCast/Float64/Float64(%93)
    output %183
  }
)
def BackProp_3094_3100(id: @3100) : (Class-Object, Class-Object, Float64) -> Unit = func(
  thunk %3097 = { env(id: %3170) : Class-Object, __auto_param_274(id: %3098) : Class-__AD__Accumulators_6_ha, __auto_param_275(id: %3099) : Float64 =>
    def %3101 : Class-__AD__Accumulators_6_ha = deref(__auto_param_274(id: %3098))
    def %3102 : Float64 = field/1(%3101)
    def %9386 : Float64 = plus(__auto_param_275(id: %3099), %3102)
    def %3104 : Unit = asg/1(%9386, __auto_param_274(id: %3098))
    output %3104
  }
)
def @3173 : Class-__Auto__Environment_20 = class/1/0(%198)
def Class-Object_tangentAdd(id: @574) : (Class-Object, Class-Object) -> Class-Object = func(
  thunk %571 = { __auto_param_48(id: %572) : Class-Object, __auto_param_49(id: %573) : Class-Object =>
    def %576 : Interface-ADEnvironment = typeCast/Class-Object/Interface-ADEnvironment(__auto_param_48(id: %572))
    def %577 : Interface-ADEnvironment = typeCast/Class-Object/Interface-ADEnvironment(__auto_param_49(id: %573))
    def %575 : Bool = instanceof(__auto_param_48(id: %572), ADEnvironment(id: %422))
    def %582 : Interface-ADEnvironment = if(%575;
      thunk %579 = {  =>
        def %581 : Interface-ADEnvironment = invoke(%576, TangentAdd(id: %580), %577)
        output %581
      }
      thunk %578 = {  =>
        output %577
      }
    )
    def %583 : Class-Object = typeCast/Interface-ADEnvironment/Class-Object(%582)
    output %583
  }
)
def Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit>_tangentAdd(id: @1661) : (Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit>, Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit>) -> Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit> = func(
  thunk %1658 = { __auto_param_142(id: %1659) : Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit>, __auto_param_143(id: %1660) : Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit> =>
    def %1662 : Class-Object = field/0(__auto_param_142(id: %1659))
    def %1663 : Class-Object = field/0(__auto_param_143(id: %1660))
    def %1665 : Class-Object = apply/2(Class-Object_tangentAdd(id: @574), %1662, %1663)
    def %1664 : (Class-Object, Class-Object, Float64) -> Unit = null
    def %1666 : Closure<Class-Object, (Class-Object, Class-Object, Float64) -> Unit> = tuple/2(%1665, %1664)
    output %1666
  }
)
output goo2(id: @61)
